File: solutions/ruby/binary-search-tree/1/binary_search_tree.rb
class Bst
  attr_reader :data, :left, :right

  def initialize(data)
    @data = data
    @left = nil
    @right = nil
  end

  # Inserta un nuevo valor en el lugar que le corresponde de forma recursiva
  def insert(new_data)
    if new_data <= @data
      # Si es menor o igual, va a la izquierda
      if @left.nil?
        @left = Bst.new(new_data)
      else
        @left.insert(new_data) # Delegación recursiva
      end
    else
      # Si es mayor, va a la derecha
      if @right.nil?
        @right = Bst.new(new_data)
      else
        @right.insert(new_data) # Delegación recursiva
      end
    end
    self
  end

  # Recorre el árbol en orden (In-Order Traversal) para devolver un arreglo ordenado
  # Si se le pasa un bloque, va entregando los datos uno a uno (comportamiento de enumerador)
  def each(&block)
    return to_enum(:each) unless block_given?

    # 1. Visitar subárbol izquierdo (valores más chicos)
    @left.each(&block) if @left

    # 2. Visitar la raíz actual
    yield @data

    # 3. Visitar subárbol derecho (valores más grandes)
    @right.each(&block) if @right
  end
end

File: solutions/typescript/binary-search-tree/1/binary-search-tree.ts
export class BinarySearchTree<T = number> {
  private _data: T;
  private _left?: BinarySearchTree<T>;
  private _right?: BinarySearchTree<T>;

  constructor(data: T) {
    this._data = data;
  }

  // Getters requeridos por la especificación de las pruebas
  public get data(): T {
    return this._data;
  }

  public get left(): BinarySearchTree<T> | undefined {
    return this._left;
  }

  public get right(): BinarySearchTree<T> | undefined {
    return this._right;
  }

  /**
   * Inserta un nuevo elemento en el lugar correcto del árbol.
   */
  public insert(item: T): void {
    if (item <= this._data) {
      // Va a la izquierda (menor o igual)
      if (this._left) {
        this._left.insert(item);
      } else {
        this._left = new BinarySearchTree(item);
      }
    } else {
      // Va a la derecha (mayor)
      if (this._right) {
        this._right.insert(item);
      } else {
        this._right = new BinarySearchTree(item);
      }
    }
  }

  /**
   * Recorre el árbol en orden (In-Order) ejecutando un callback por cada nodo.
   * Esto garantiza procesar los datos de manera completamente ordenada.
   */
  public each(callback: (data: T) => void): void {
    // 1. Visitar de forma recursiva el subárbol izquierdo
    if (this._left) {
      this._left.each(callback);
    }

    // 2. Procesar el nodo actual
    callback(this._data);

    // 3. Visitar de forma recursiva el subárbol derecho
    if (this._right) {
      this._right.each(callback);
    }
  }
}